### PR TITLE
[WIP] T5v1.1 & MT5

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by shenfe** [comment]
Maybe wrong model config for T5.1.1. For instance, T5.1.1.small should have num_layers=8 and num_heads=6.

See https://github.com/google-research/text-to-text-transfer-transformer/blob/master/t5/models/gin/models/t5.1.1.small.gin

**Review by shenfe — CHANGES_REQUESTED** [review]
Hi, there are some problems I found:
- different model configs, compared with the [official](https://github.com/google-research/text-to-text-transfer-transformer/tree/master/t5/models/gin/models)
- relu maybe unnecessary in FFN
- lm_head weight should not share with the embedding of enc/dec

And just point it out if anything wrong :)

**Comment by patrickvonplaten** [comment]
> Maybe wrong model config for T5.1.1. For instance, T5.1.1.small should have num_layers=8 and num_heads=6.
> 
> See https://github.com/google-research/text-to-text-transfer-transformer/blob/master/t5/models/gin/models/t5.1.1.small.gin

Thanks yeah, I implemented that. 

The new model structure is now equal to mesh t5 v1.1. 

If you download the t5v1.1 `t5-small` checkpoint and replace the corresponding path in `check_t5_against_hf.py` you can see that the models are equal. 

There is still quite some work to do: write more tests, lots of cleaning and better design, and check if mT5 works with it.

**Comment by shenfe** [comment]
> If you download the t5v1.1 `t5-small` checkpoint and replace the corresponding path in `check_t5_against_hf.py` you can see that the models are equal.

Hi, `check_t5_against_hf.py` still fails if I use a longer input text instead of `Hello there`, like `Hello there. Let's put more words in more languages than I originally thought.`

File: check_t5_against_hf.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+model = T5v2ForConditionalGeneration.from_pretrained(path_to_tf_checkpoint, return_dict=True)
+
+input_ids = tok("Hello there", return_tensors="pt").input_ids
+labels = tok("Hi I am", return_tensors="pt").input_ids
+
+# input_ids and labels are ok!
+loss = model(input_ids, labels=labels).loss
+mesh_tf_loss = -(labels.shape[-1] * loss.item())
+
+if mesh_tf_loss - score[0][0] < 1e-4:
```
Comment (shenfe) [inline check_t5_against_hf.py:60]: Maybe better to use `abs()` here

Comment (patrickvonplaten) [reply check_t5_against_hf.py:60]: I will delete this file eventually - it's just for now :-)

**Comment by patrickvonplaten** [comment]
> > If you download the t5v1.1 `t5-small` checkpoint and replace the corresponding path in `check_t5_against_hf.py` you can see that the models are equal.
> 
> Hi, `check_t5_against_hf.py` still fails if I use a longer input text instead of `Hello there`, like `Hello there. Let's put more words in more languages than I originally thought.`

Hmm, it works for me - do you experience that for T5v1.1 or mT5?

File: src/transformers/modeling_t5v2.py [hunk 2]
```diff
@@ -0,0 +1,1312 @@
...
+        # layer norm should always be calculated in float32
+        variance = x.to(torch.float32).pow(2).mean(-1, keepdim=True)
+        x = x / torch.sqrt(variance + self.variance_epsilon)
+
+        if self.weight.dtype == torch.float16:
+            x = x.to(torch.float16)
+        return self.weight * x
+
+
+class T5v2DenseReluDense(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_t5v2.py:196]: This class differs from previous `T5DenseReluDense` significantly -> different weights are used here

File: src/transformers/modeling_t5v2.py [hunk 3]
```diff
@@ -0,0 +1,1312 @@
...
+            encoder_attention_mask=attention_mask,
+            head_mask=head_mask,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        sequence_output = decoder_outputs[0]
+        # Rescale output before projecting on vocab
```
Comment (patrickvonplaten) [inline src/transformers/modeling_t5v2.py:1248]: For T5v1.1 there is no rescaling because the input and output embedding are **not** tied.

**Comment by shenfe** [comment]
@patrickvonplaten yes, I use the latest `check_t5_against_hf .py` script and mT5.

I mean, `if mesh_tf_loss - score[0][0] < 1e-4:` is not enough, because actually the two values are -84.91265869140625 and -64.86555480957031. It would fail then if you use `abs()`.

```py
if abs(mesh_tf_loss - score[0][0]) < 1e-4:  # <= use `abs`
    print("Success!")
else:
    print(f"Fail. Mesh TF {mesh_tf_loss} vs. {score[0][0]}")
```

```
Fail. Mesh TF -84.91265869140625 vs. -64.86555480957031
```

File: src/transformers/modeling_t5v2.py (hunk 3 above)
Comment (shenfe) [reply src/transformers/modeling_t5v2.py:1248]: I agree on these changes. But the PyTorch T5.1.1 model still differs from the official tf version. I'm working on it too.

**Comment by shenfe** [comment]
> > > If you download the t5v1.1 `t5-small` checkpoint and replace the corresponding path in `check_t5_against_hf.py` you can see that the models are equal.
> > 
> > 
> > Hi, `check_t5_against_hf.py` still fails if I use a longer input text instead of `Hello there`, like `Hello there. Let's put more words in more languages than I originally thought.`
> 
> Hmm, it works for me - do you experience that for T5v1.1 or mT5?

Aha, the checking is OK now. Yesterday I made a mistake that when I changed the test input sentence in the check script, I didn't update the input length for MTF model from 4 to a longer value like 128. So actually the MTF model and PyTorch model received different inputs, and of course got different results.

Besides, if I add the z-loss to the CE loss at last, it differs from MTF score again. I just found MTF ignores z-loss when not training ([code](https://github.com/tensorflow/mesh/blob/4f82ba1275e4c335348019fee7974d11ac0c9649/mesh_tensorflow/transformer/transformer.py#L781)). So I think MTF model score does not include z-loss, but its training does, which is absent from HF T5 training. Well, this is absolutely not a blocking issue now.

Appreciate your great work :)

**Comment by patrickvonplaten** [comment]
closing in favor of https://github.com/huggingface/transformers/pull/8552.
